### DISCUSSION TITLE
How to validate query parameters

### DISCUSSION THREAD

**Discussion by tiaanl** [rwf2/Rocket#1988] [Questions]
Let's use the oauth2 authorize endpoint for example:

```
/oauth2/authorize?<client_id>&<redirect_uri>&<code_challenge>&<code_challenge_method>
```

I want to have validation and "feedback" from those query parameters, e.g. if the `client_id` field is not given I want to return a bad request (400) with a message "client_id is required".  Same with `redirect_uri`.

Another use case would be for `code_challenge_method`, e.g.  If the value is not equal to `S256` I would like to return a bad request (400) stating "code_challenge_method must be S256".

I cannot find a way using forms to get the desired effect, because if the form fails validation, it fails to match the route and returns 404. Does this mean I have to pass in everything as `Option<_>` and handle all the validation manually?

**Comment by tobymurray** [comment]
I think you should be able to do this with a [request guard](https://rocket.rs/v0.5-rc/guide/requests/#request-guards)? If you don't need access to the request body, you can check whatever you'd like from the query string and provide the corresponding errors.

Not a full snippet, but your `from_request` could include things like:
```
let query_string = match request.uri().query() {
	None => return Outcome::Failure((Status::BadRequest, YourError::NoQueryString)),
	Some(q) => q,
};

let client_id = match query_string.segments().find(|&param| param.0 == "client_id") {
	None => return Outcome::Failure((Status::BadRequest, YourError::QueryStringButNoClientId)),
	Some(h) => h,
};
```

Or what have you. That'd give you control over the specific status for each condition.
